### ISSUE TITLE
Getting error when trying to calculate the flux after a successful fit

### ISSUE THREAD

**Issue by ivvv** [threeML/threeML#542]
**Describe the bug**

I successfully fit a model to a spectrum of a galaxy cluster observed with XMM, EPIC-pn. The fit results look as expected. Trying to calculate the flux in a certain energy range, however, result in an attribute error:

```
AttributeError: Accessing an element session of the node that does not exist
```

I've put an example jupyterlab notebook and the necessary spectral files for testing it in [this zip archive](https://www.dropbox.com/s/5ei25yyrrk9ti30/a0133_3ml_test.zip?dl=0).

I could not figure out what I was doing wrong, tried with different inputs to `get_flux()` but no success.

Any ideas?

**Comment by grburgess** [comment]
Hi, I think I have an idea. The APEC model is out of date with the newer framework. I will have a look in a few hours.

**Comment by grburgess** [comment]
Indeed, this will be fixed in the upcoming dev version.

**Comment by grburgess** [comment]
@ivvv Ok, sorry for the trouble. If you can try to install astromodels from the dev version.

The issue was the abundance table was set in the old way that we utilized external functions and it does not regenerate for some things. The fix should have it. Now instead of doing xxxx.init_session() 

do: xxxx.abundance_table = "<pytomdb abundance>" 

It should default to 'AG89'

**Comment by ivvv** [comment]
Right,just  git cloned the latest astromodels and installed it with no errors (some tweaks with the XSPEC version). Import of astromodels works and reports the correct version: 
```
In [4]: astromodels.__version__
Out[4]: '2.3.2+0.gbfad699.dirty'
```

And then in my code/notebook commented out the old call and added your suggestion:

```
modapec = APEC()
#modapec.init_session(abund_table='AG89')
modapec.abund_table='AG89'
```
And now the code fails at the fitting stage: `jl.fit()` with the same error:

`AttributeError: Accessing an element session of the node that does not exist`

**Comment by grburgess** [comment]
Thanks! I'll keep working to figure this out

**Comment by grburgess** [comment]
Ok, I fixed the other issue with not finding apec... this was a bug that creeped in. 

I just tested locally and everything should work now. Can you try to update and check from the dev branch? 

Make sure to pip uninstall astromodels (even a few times).

**Comment by grburgess** [comment]
@ivvv I have been trying to understand the pyatomdb code to allow one to input single energies for a computation, but I have yet to be able to figure this out. @domeckert would you have any idea how to do this? I suppose it is something in the set_response function.

**Comment by domeckert** [comment]
Hi @grburgess  and @ivvv . First of all it is nice to see that the code is being used! 

I think you are right that the issue comes from using single values rather than arrays. To check whether that is the case, can you try to pass an array including the value of interest rather than the value itself? Something like

`elow = np.array([0.5*u.keV])`
`ehigh = np.array([2.0*u.keV])`
`.get_flux(0.5*u.keV,2.0*u.keV)`

**Comment by grburgess** [comment]
@domeckert I think it is in the integration code for the flux itself. Depending on the method, scipy will pass either an array or a single value to the function. Currently, the APEC function cannot accept single values. I tried to fix that myself, but I failed and I do not know enough about pyatomdb to modify that part. The line is here: https://github.com/threeML/astromodels/blob/dev/astromodels/functions/functions_1D/apec.py#L122

**Comment by ivvv** [comment]
Yesterday I tried what @domeckert suggested, and also different ways to pass arguments, as lists, numpy arrays, with and without unites etc. No success up to now.

**Comment by domeckert** [comment]
OK. I think we can get around this issue by introducing a try statement that will check whether the input value is a scalar or an array.

**Comment by ivvv** [comment]
Pulled the dev version of `threeml`, version is `2.3.1+40.g2ee7bfb8` and the same error as in previous message (attachment). 

I called the method with `.get_flux(0.5*u.keV,2.0*u.keV)`.

I tried different ways to provide the emin and emax, unitless, with eV units, as arrays, and no luck.

**Comment by ivvv** [comment]
To help with debugging, I saved the results in a fits file with `jl.results.write_to()` method. The results can be read with `load_analysis_results()`  and used for `get_flux()`. Here is a short code:

```
from threeML import *
import astropy.units as u

qres = load_analysis_results('3ml_results_save.fits.gz')
qres.display('hpd')
elow = 0.5*u.keV
ehigh = 2.0*u.keV
qres.get_flux(elow,ehigh)
```

I tried with very simple modifications in source code, function `apec.py` but nothing really worked... e.g like this one in method `evaluate()`:

```
            ebounds = np.empty(nval + 1)

# Ivan's modif

            if isinstance(x,astropy_units.quantity.Quantity):

                ebounds = ebounds * x.unit

```

[3ml_results_save.fits.gz](https://github.com/threeML/threeML/files/9218825/3ml_results_save.fits.gz)

**Comment by henrikef** [comment]
have you tried
https://numpy.org/doc/stable/reference/generated/numpy.atleast_1d.html ?

**Comment by grburgess** [comment]
Yeah, the problem is that it is how the single numbers interact with the ATOMDB fake response. We need someone that understands that. Or I need to make an alternative flux computer.

**Comment by domeckert** [comment]
I can't claim I understand  that very well but I can certainly have a look.

**Comment by grburgess** [comment]
maybe @rwiller can also assist as he has been using the APEC model a bit.
